src/uext.cpp

```diff
@@ -0,0 +1,75 @@
+/* LibreSolar charge controller firmware
+ * Copyright (c) 2016-2019 Martin Jäger (www.libre.solar)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "uext.h"
+#include <algorithm>
+
+// we use code to self register objects at construction time.
+// It relies on the fact that initially this pointer will be initialized
+// with NULL ! This is only true for global variables or static members of classes
+// so this must remain a static class member
+std::vector<UExtInterface*>* UExtInterfaceManager::interfaces;
+
+// run the respective function on all objects in the "devices" list
+// use the c++11 lambda expressions here for the for_each loop, keeps things compact
+
+void UExtInterfaceManager::process_asap()
+{
+    for_each(std::begin(*interfaces),std::end(*interfaces), [](UExtInterface* tsif) {
+        tsif->process_asap();
+    });
+}
+
+void UExtInterfaceManager::enable()
+{
+    for_each(std::begin(*interfaces),std::end(*interfaces), [](UExtInterface* tsif) {
+        tsif->enable();
+    });
+}
+
+void UExtInterfaceManager::process_1s()
+{
+    for_each(std::begin(*interfaces),std::end(*interfaces), [](UExtInterface* tsif) {
+        tsif->process_1s();
+    });
+}
+
+void UExtInterfaceManager::checkList()
+{
+    if (UExtInterfaceManager::interfaces == NULL)
+    {
+        UExtInterfaceManager::interfaces = new std::vector<UExtInterface*>;
```

Comment:
With the new statement we start using the heap. So far, I always strictly avoided any dynamic memory allocation. But as far as I can see this is only called once during initialization, so should not create issues, correct?

Does the C++ vector use dynamic memory allocation internally? If not, how could it be extended dynamically (using `push_back` function? And does this add much overhead? I'm always a bit cautious when using advanced C++ features for embedded programs.

Reply:
Yes, dynamic allocation is a dangerous thing if used without caution. However, as you pointed out, we use it only in limited capacity and only additively. We are talking a couple of bytes here. There is no real heap fragmentation and not a lot of overhead in terms of memory usage. 
At runtime the overhead is there but we are talking again only a couple of instructions (iterate through list, for each item load pointer to object, load pointer to objects virtual function table, load pointer from virtual function table in object, call code at this location).

This is the price to pay for flexibility and no code changes at the same time. However, I would venture to say that a single floating point div takes longer than this code (since we are having no FPU). One always has to see this in relation to rest of the program.

BTW, yes vector has to allocate space for items dynamically. With some extra work we could avoid of this. Not sure if this pays off in our use case.

Reply:
Ok, thanks for the explanation. I agree that it's fine to use dynamic allocation here.

Regarding floating points: Yes, that was a decision I made some time ago. After having some bugs because of variable overflows or rounding errors because of wrong order of operations I thought that loosing a bit of performance (or increasing energy consumption a bit, as the processor is powerful enough, but just takes longer) is worth the much lower probability of bugs if you don't need to take care of the decimal point, correct units and overruns.